File: backend/app/api/routes/opml.py
from __future__ import annotations

import xml.etree.ElementTree as ET
from datetime import datetime, timezone
from io import StringIO

from fastapi import APIRouter, Depends, File, HTTPException, Response, UploadFile
from sqlmodel import Session, select

from app.db.deps import get_session
from app.db.models import Feed

router = APIRouter(prefix="/opml", tags=["opml"])


@router.get("/export")
async def export_opml(session: Session = Depends(get_session)) -> Response:
    """Export all feeds as OPML."""
    feeds = session.exec(select(Feed)).all()
    
    # Create OPML XML structure
    opml = ET.Element("opml", version="2.0")
    head = ET.SubElement(opml, "head")
    
    title = ET.SubElement(head, "title")
    title.text = "RSS READER Subscriptions"
    
    date_created = ET.SubElement(head, "dateCreated")
    date_created.text = datetime.now(timezone.utc).strftime("%a, %d %b %Y %H:%M:%S GMT")
    
    body = ET.SubElement(opml, "body")
    
    # Group feeds by group_name
    groups: dict[str, list[Feed]] = {}
    for feed in feeds:
        group_name = feed.group_name or "default"
        if group_name not in groups:
            groups[group_name] = []
        groups[group_name].append(feed)
    
    # Create outline elements
    for group_name, group_feeds in groups.items():
        if len(groups) > 1:
            # Create group outline
            group_outline = ET.SubElement(body, "outline", text=group_name, title=group_name)
            for feed in group_feeds:
                ET.SubElement(
                    group_outline,
                    "outline",
                    type="rss",
                    text=feed.title or feed.url,
                    title=feed.title or feed.url,
                    xmlUrl=feed.url,
                    htmlUrl=feed.site_url or "",
                )
        else:
            # No grouping, add feeds directly
            for feed in group_feeds:
                ET.SubElement(
                    body,
                    "outline",
                    type="rss",
                    text=feed.title or feed.url,
                    title=feed.title or feed.url,
                    xmlUrl=feed.url,
                    htmlUrl=feed.site_url or "",
                )
    
    # Convert to string
    tree = ET.ElementTree(opml)
    output = StringIO()
    tree.write(output, encoding="unicode", xml_declaration=True)
    
    return Response(
        content=output.getvalue(),
        media_type="application/xml",
        headers={"Content-Disposition": "attachment; filename=rss_subscriptions.opml"},
    )


@router.post("/import")
async def import_opml(
    file: UploadFile = File(...), session: Session = Depends(get_session)
) -> dict[str, int | list[str]]:
    """Import feeds from OPML file."""
    if not file.filename or not file.filename.endswith((".opml", ".xml")):
        raise HTTPException(status_code=400, detail="Invalid file type. Please upload an OPML file.")
    
    content = await file.read()
    
    try:
        root = ET.fromstring(content.decode("utf-8"))
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Failed to parse OPML: {str(e)}")
    
    # Find all outline elements with type="rss" or xmlUrl attribute
    imported = 0
    skipped = 0
    errors: list[str] = []
    
    def process_outline(outline: ET.Element, parent_title: str | None = None) -> None:
        nonlocal imported, skipped
        
        xml_url = outline.get("xmlUrl")
        if xml_url:
            # This is a feed
            title = outline.get("title") or outline.get("text") or xml_url
            group_name = parent_title or "default"
            
            # Check if feed already exists
            try:
                existing = session.exec(select(Feed).where(Feed.url == xml_url)).first()
                if existing:
                    skipped += 1
                    return
                
                # Add new feed
                feed = Feed(
                    url=xml_url,
                    title=title,
                    site_url=outline.get("htmlUrl") or "",
                    group_name=group_name,
                )
                session.add(feed)
                imported += 1
            except Exception as e:
                errors.append(f"Error adding feed {xml_url}: {str(e)}")
        
        # Process children regardless (some feeds might have both xmlUrl and children)
        children = outline.findall("outline")
        if children:
            current_title = outline.get("title") or outline.get("text") or parent_title
            for child in children:
                process_outline(child, current_title)
    
    # Process body
    body = root.find("body")
    if body is not None:
        for outline in body.findall("outline"):
            try:
                process_outline(outline)
            except Exception as e:
                errors.append(f"Error processing outline: {str(e)}")
    
    session.commit()
    
    return {
        "imported": imported,
        "skipped": skipped,
        "errors": errors,
    }

File: backend-node/src/routes/opml.ts
/**
 * OPML API Routes (Import/Export)
 */

import { FastifyInstance } from 'fastify';
import { FeedRepository } from '../db/repositories/index.js';
import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import { refreshFeed } from '../services/fetcher.js';

export async function opmlRoutes(app: FastifyInstance) {
  const feedRepo = new FeedRepository();

  // POST /opml/import - Import OPML file
  app.post('/opml/import', async (request, reply) => {
    try {
      // Handle multipart file upload
      const data = await request.file();

      if (!data) {
        return reply.code(400).send({ error: 'No file uploaded' });
      }

      // Read file content
      const buffer = await data.toBuffer();
      const opml_content = buffer.toString('utf-8');

      // Get group_name from fields if provided
      const group_name = data.fields.group_name?.value as string | undefined;

      if (!opml_content) {
        return reply.code(400).send({ error: 'OPML content is required' });
      }

      const parser = new XMLParser({
        ignoreAttributes: false,
        attributeNamePrefix: '@_',
      });

      const result = parser.parse(opml_content);

      if (!result.opml || !result.opml.body) {
        return reply.code(400).send({ error: 'Invalid OPML format' });
      }

      const imported: any[] = [];
      const skipped: any[] = [];
      const errors: string[] = [];

      // Process outline elements (feeds)
      const processOutline = (outline: any, parentGroup?: string) => {
        if (!outline) return;

        // Handle array of outlines
        const outlines = Array.isArray(outline) ? outline : [outline];

        for (const item of outlines) {
          // Check if this is a folder/category (has nested outlines)
          if (item.outline) {
            const folderName = item['@_text'] || item['@_title'] || 'Uncategorized';
            processOutline(item.outline, folderName);
            continue;
          }

          // This is a feed
          const feedUrl = item['@_xmlUrl'] || item['@_url'];
          const feedTitle = item['@_text'] || item['@_title'];
          const feedGroup = parentGroup || group_name || 'default';

          if (!feedUrl) {
            continue;
          }

          try {
            // Check if feed already exists
            const existing = feedRepo.findByUrl(feedUrl);
            if (existing) {
              skipped.push({
                url: feedUrl,
                title: feedTitle,
                reason: 'Feed already exists',
              });
              continue;
            }

            // Create new feed
            const feed = feedRepo.create({
              url: feedUrl,
              title: feedTitle || null,
              group_name: feedGroup,
            });

            imported.push({
              id: feed.id,
              url: feed.url,
              title: feed.title,
              group_name: feed.group_name,
            });

            refreshFeed(feed.id).catch((error) => {
              console.error('Failed to refresh imported feed:', error);
            });
          } catch (error) {
            errors.push(error instanceof Error ? error.message : 'Unknown error');
          }
        }
      };

      processOutline(result.opml.body.outline);

      return {
        imported: imported.length,
        skipped: skipped.length,
        errors,
      };
    } catch (error) {
      return reply.code(400).send({
        error: error instanceof Error ? error.message : 'Failed to parse OPML',
      });
    }
  });

  // GET /opml/export - Export feeds to OPML
  app.get('/opml/export', async (request, reply) => {
    try {
      const feeds = feedRepo.findAll();

      // Group feeds by group_name
      const groupedFeeds: Record<string, any[]> = {};
      for (const feed of feeds) {
        const group = feed.group_name || 'default';
        if (!groupedFeeds[group]) {
          groupedFeeds[group] = [];
        }
        groupedFeeds[group].push(feed);
      }

      // Build OPML structure
      const outlines: any[] = [];

      for (const [groupName, groupFeeds] of Object.entries(groupedFeeds)) {
        if (groupName === 'default' && Object.keys(groupedFeeds).length === 1) {
          // If only default group, don't create a folder
          for (const feed of groupFeeds) {
            outlines.push({
              '@_type': 'rss',
              '@_text': feed.title || feed.url,
              '@_title': feed.title || feed.url,
              '@_xmlUrl': feed.url,
              '@_htmlUrl': feed.url,
            });
          }
        } else {
          // Create folder for group
          const folderOutlines = groupFeeds.map(feed => ({
            '@_type': 'rss',
            '@_text': feed.title || feed.url,
            '@_title': feed.title || feed.url,
            '@_xmlUrl': feed.url,
            '@_htmlUrl': feed.url,
          }));

          outlines.push({
            '@_text': groupName,
            '@_title': groupName,
            outline: folderOutlines,
          });
        }
      }

      const opmlData = {
        '?xml': {
          '@_version': '1.0',
          '@_encoding': 'UTF-8',
        },
        opml: {
          '@_version': '2.0',
          head: {
            title: 'Aurora RSS Reader Feeds',
            dateCreated: new Date().toUTCString(),
          },
          body: {
            outline: outlines,
          },
        },
      };

      const builder = new XMLBuilder({
        ignoreAttributes: false,
        attributeNamePrefix: '@_',
        format: true,
      });

      const opmlXml = builder.build(opmlData);

      reply.header('Content-Type', 'application/xml');
      reply.header('Content-Disposition', 'attachment; filename=rss_subscriptions.opml');
      return reply.send(opmlXml);
    } catch (error) {
      return reply.code(500).send({
        error: error instanceof Error ? error.message : 'Failed to export OPML',
      });
    }
  });
}
